### PR TITLE
UCD fetcher / mapper

### PR DESCRIPTION
**Pull request by lthurston**
Included:

- [x] Resolve issue with calisphere-id in UCLDC_map()
- [x] Resolve issue with legacy couch db id
- [x] Decide how to render calisphere id when missing the ark
- [x] Properly write to local or s3 after recent refactor
- [x] Add type annotations
- [x] Create a general function for fetching URLs w/ retries (`make_http_request()`) in the Fetcher base class

### CODE REVIEW

**Comment by lthurston** [comment]
Docker changes have been removed from this branch.

**Comment by christinklez** [comment]
Hi @lthurston! We've reviewed the validation reports for the UCD fetcher/mapper set.
Only linking to one particular validation report, for [Collection #8](https://docs.google.com/spreadsheets/d/1WtWtx1FYlCuosQei-eyLsCckMdrL2AiGcU8-VYq7hQI/edit#gid=1478019524), but the issues listed below showed up in many other collections. (Let me know if you'd like to see the other reports, but note that the linked #8 report has the most bit of annotated notes for these issues. Thank you!

- **title & description:** We expect the data to have brackets wrapping around the text; Rikolti is not wrapping brackets. (Is this a mapping issue?) See examples in rows 2 & 7. Please wrap the data in brackets
- **rights / rights_uri:** We expect the rights info to map to rights_uri; it is currently mapping to rights. Please map to rights_uri. See examples in rows 4 & 6. 
- **title & rights -- type mismatch:** Type mismatch errors--expect "list of str" vs actual "str". Please investigate type mismatch errors.
- **subject:** Subject terms are not mapping. Subject terms seem to appear in the "about" fields in the vernacular data. See vernacular example in row 8. Please map subject terms.
- **publisher:** 20 records in this collection are incorrectly mapping the publisher data. mapped metadata is showing: "publisher": ["name"]. We expect to see what's currently in the expected value ("UCD Library..."). See example (as well as vernacular data variations) in row 15653. Please investigate publisher mapping anomaly (in 20 records only; other records are okay.)

**Comment by lthurston** [comment]
@christinklez I've addressed these issues in this PR.

File: metadata_fetcher/fetchers/ucd_json_fetcher.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            params: dict[str]
+        """
+        super(UcdJsonFetcher, self).__init__(params)
+
+        self.collection_id = params.get("collection_id")
+        self.url = params.get("harvest_data").get("url")
+        self.per_page = 10
+
+    def fetch_page(self) -> int:
+        """
```
Comment (amywieliczka) [inline metadata_fetcher/fetchers/ucd_json_fetcher.py:24]:
```suggestion
        """
        UCD's harvesting endpoint gets us an XML document listing a URL for every record in a collection, but not the actual metadata records themselves. fetch_page fetches this initial XML document, while fetch_all_pages below batches the records in sets of 10 (self.per_page) and fetch_json_ld fetches individual records.
```

File: metadata_mapper/lambda_function.py [hunk 2]
```diff
@@ -102,13 +102,13 @@ def map_page(collection_id: int, page_filename: str, collection: Union[dict, str
     mapped_records = [record.solr_updater() for record in mapped_records]
     mapped_records = [record.remove_none_values() for record in mapped_records]
 
+    group_page_exceptions = {}
```
Comment (amywieliczka) [inline metadata_mapper/lambda_function.py:105]: Thank you

**Review by amywieliczka — APPROVED** [review]
Looks good to me. Just added a code comment suggestion

**Pull request merged**